src/core/unit.c

```diff
@@ -319,6 +352,11 @@ int unit_choose_id(Unit *u, const char *name) {
                 name = t;
         }
 
+        if (streq(u->id, name))
```

Comment:
initially Unit structures have no name, and hence no id. Should be streq_ptr(), no?

Reply:
But unit_choose_id() is not called on them? unit_choose_id is called from merge_by_names, in which case both sides should have .id set, and from socket_enter_running(), which is requesting an existing unit from the manager.

Reply:
maybe make this explicit then? i.e. `assert(u->id); /* this function is only called when we already have an id */`

Reply:
(that said, I'd keep the function generic and not encode too many expectations needless, i.e. fail cleanly if u->id is not set at all yet, or make it handle that. i mean, who knows what code will eventually end up calling this function)

Reply:
It is now reworked to work fine if u->id is not set.